optimum/amd/brevitas/quantizer.py

```diff
@@ -0,0 +1,357 @@
+                )
+
+            with torch.no_grad():
+                model = symbolic_trace(self.model, input_names)
+        else:
+            model = self.model
+
+        # Because accelerate is not compatible with FX, we keep two versions of the Model
+        # one with FX-traced, the other one not.
+        # Since weights are shared across the two, we can apply weight/activation equalization
+        # by using one representation or the other based on needs.
+        if quantization_config.apply_weight_equalization:
+            logger.info("Applying weight equalization...")
+            apply_weight_equalization(model)
+            logger.info("Weight equalization applied.")
+
+        if quantization_config.activations_equalization is not None:
+            logger.info(
+                f"Applying Activation Equalization {quantization_config.activations_equalization} (SmoothQuant)..."
+            )
+            apply_act_equalization(model, quantization_config.activations_equalization, calibration_dataset)
+            logger.info("Activation equalization applied.")
+
+        if not use_accelerate:
+            context = torch.device(next(model.parameters()).device)
+        else:
+            context = contextlib.nullcontext()
+
+        if use_accelerate:
+            remove_hooks(model)
+
+        # We do not quantize embedding and last fully connected layer
+        with context:
+            model = quantize_model(
+                model,
+                dtype=dtype,
+                weight_quant_format="int",
+                weight_quant_type="sym" if quantization_config.weights_symmetric else "asym",
+                weight_bit_width=quantization_config.weights_bitwidth,
+                weight_param_method=quantization_config.weights_param_method,
+                weight_scale_precision=quantization_config.scale_precision,
+                weight_quant_granularity=quantization_config.weights_quant_granularity,
+                weight_group_size=quantization_config.weights_group_size,
+                quantize_weight_zero_point=quantization_config.quantize_zero_point,
+                input_bit_width=quantization_config.activations_bitwidth,
+                input_quant_type="sym" if quantization_config.activations_symmetric else "asym",
+                input_quant_format="int",
+                input_param_method=quantization_config.activations_param_method,
+                input_scale_precision=quantization_config.scale_precision,
+                input_scale_type="static" if quantization_config.is_static else "dynamic",
+                input_quant_granularity=quantization_config.activations_quant_granularity,
+                input_group_size=quantization_config.activations_group_size,
+                quantize_input_zero_point=quantization_config.quantize_zero_point,
+            )
+
+        if use_accelerate:
```

Comment:
This has to go after the call to the forward (current lines 207-211)

Reply:
@Giuseppe5 Why? Calling `offload_model` after the forward call results in the call `model(**calibration_dataset[0])` being executed on CPU (in case accelerate is used).

Reply:
Many quantization parameters do not exist before that first forward call, so accelerate is not aware of them (and therefore can't manage them)

Reply:
To add one more bit, unless a forward pass has happened (to trigger the computation of scale factors), when calling `state_dict`, we explicitly remove the scale factor since its value would be incorrect.
Since `offload_model` relies on a correctly formed `state_dict`, we need to have a CPU-only forward pass.